metagraph/tests/test_vector_algorithm.cpp

```diff
@@ -123,7 +123,162 @@ TEST(IntVector, call_nonzeros_sparse) {
+                uint64_t it = begin;
+                call_nonzeros(vector, begin, end, [&](auto i, auto count) {
+                    EXPECT_LE(begin, i);
+                    EXPECT_GT(end, i);
+                    while (vector[it] == 0) {
+                        EXPECT_GT(i, it);
+                        ++it;
+                        ASSERT_GT(end, it);
+                    }
+
+                    EXPECT_EQ(vector[i], count);
+                    EXPECT_EQ(it++, i);
+                });
+                for (; it < end; ++it) {
+                    EXPECT_EQ(0u, vector[it]);
+                }
+            }
+        }
+    }
+}
+
 TEST(IntVector, call_nonzeros_sparse_every_4) {
+    for (size_t w = 1; w <= 64; ++w) {
+        sdsl::int_vector<> vector_relaxed = aligned_int_vector(1000, 0, w, 16);
+        sdsl::int_vector<> vector_seq_cst = aligned_int_vector(1000, 0, w, 16);
+        sdsl::int_vector<> vector(1000, 0, w);
+        ASSERT_LT(65u, vector.size());
+        std::atomic<size_t> counter = 0;
+
+        std::mutex mu;
+        std::mutex mu2;
+        #pragma omp parallel for num_threads(3) schedule(dynamic)
+        for (size_t j = 0; j < 3; ++j) {
+            for (size_t i = j; i < vector.size(); i += 3) {
+                if (i % 130 == 1) {
+                    uint64_t val = (++counter) % w;
+                    EXPECT_EQ(0u, atomic_exchange(vector_relaxed, i, val, mu, __ATOMIC_RELAXED));
+                    EXPECT_EQ(0u, atomic_exchange(vector_seq_cst, i, val, mu));
+                    std::lock_guard<std::mutex> lock(mu2);
+                    vector[i] = val;
+                }
+            }
+        }
+
+        for (size_t i = 0; i < vector.size(); ++i) {
+            EXPECT_EQ(vector[i], atomic_fetch(vector_relaxed, i, mu, __ATOMIC_ACQUIRE));
+            EXPECT_EQ(vector[i], atomic_fetch(vector_seq_cst, i, mu));
+        }
+
+        for (size_t begin = 0; begin < vector.size(); begin += vector.size() / 4) {
+            for (size_t end = begin; end < vector.size(); ++end) {
+                uint64_t it = begin;
+                call_nonzeros(vector, begin, end, [&](auto i, auto count) {
+                    EXPECT_LE(begin, i);
+                    EXPECT_GT(end, i);
+                    while (vector[it] == 0) {
+                        EXPECT_GT(i, it);
+                        ++it;
+                        ASSERT_GT(end, it);
+                    }
+
+                    EXPECT_EQ(vector[i], count);
+                    EXPECT_EQ(it++, i);
+                });
+                for (; it < end; ++it) {
+                    EXPECT_EQ(0u, vector[it]) << vector.size()
+                                              << "\n" << begin
+                                              << " " << end
+                                              << "\n" << w
+                                              << "\n" << vector;
+                }
+            }
+        }
+    }
+}
```

Comment:
seems like a huge test. Hard to see what you want to check here.
Try to break it down into two or more tests and test simple things there.

Otherwise, describe in a comment why you need such a complex test.

Reply:
I reorganized these tests to get rid of the redundant checks